File: frontend/src/stores/invoice.ts
import { defineStore } from 'pinia'
import {
  getInvoicesApi,
  getInvoiceByIDApi,
  createInvoiceApi,
  updateInvoiceApi,
  updateInvoiceStatusApi,
  deleteInvoiceApi,
  cancelInvoiceApi,
  sendInvoiceApi,
  bulkSendInvoiceApi,
} from '@/api/invoice'

export interface InvoiceItem {
  id: string
  company_id: string
  invoice_id: string
  product_id: string | null
  description: string
  quantity: string
  unit: string
  unit_price: string
  discount_rate: string
  tax_rate: string
  line_total: string
  currency: string
  exchange_rate: string
  exchange_rate_op: '*' | '/'
}

export interface UserRef {
  id: string
  name: string
  role: string
}

export interface Invoice {
  id: string
  company_id: string
  cari_id: string
  type: 'sales' | 'purchase'
  number: string
  date: string
  due_date: string
  currency: string
  exchange_rate: string
  subtotal: string
  discount_total: string
  tax_total: string
  total: string
  paid_total: string
  status: 'draft' | 'sent' | 'disputed' | 'partial' | 'paid' | 'canceled'
  process_stock?: boolean
  note: string
  dispute_note?: string
  disputed_at?: string
  sent_at?: string
  last_sent_at?: string
  items?: InvoiceItem[]
  created_at: string
  updated_at: string
  created_by_user?: UserRef | null
  updated_by_user?: UserRef | null
}

export const useInvoiceStore = defineStore('invoice', {
  state: () => ({
    invoices: [] as Invoice[],
    total: 0,
    activeInvoice: null as Invoice | null,
    loading: false,
  }),
  actions: {
    async fetchInvoices(params?: any) {
      this.loading = true
      try {
        const res = await getInvoicesApi(params)
        this.invoices = res.data.data
        this.total = res.data.meta.total
      } finally {
        this.loading = false
      }
    },
    async fetchInvoiceByID(id: string) {
      this.loading = true
      try {
        const res = await getInvoiceByIDApi(id)
        this.activeInvoice = res.data.data
      } finally {
        this.loading = false
      }
    },
    async createInvoice(data: any) {
      const res = await createInvoiceApi(data)
      return res.data.data
    },
    async updateInvoice(id: string, data: any) {
      const res = await updateInvoiceApi(id, data)
      return res.data.data
    },
    async updateInvoiceStatus(id: string, status: string, paidTotal: number) {
      const res = await updateInvoiceStatusApi(id, status, paidTotal)
      return res.data.data
    },
    async deleteInvoice(id: string) {
      await deleteInvoiceApi(id)
    },
    async cancelInvoice(id: string) {
      await cancelInvoiceApi(id)
    },
    async sendInvoice(id: string) {
      await sendInvoiceApi(id)
    },
    async bulkSendInvoice(ids: string[]) {
      const res = await bulkSendInvoiceApi(ids)
      return res.data.data as { sent: string[]; failed: { id: string; error: string }[] }
    },
  },
})

File: backend/models/invoice.go
package models

import (
	"github.com/shopspring/decimal"

	"time"

	"github.com/google/uuid"
)

type Invoice struct {
	ID            uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID     uuid.UUID       `gorm:"type:uuid;not null;index" json:"company_id"`
	CariID        uuid.UUID       `gorm:"type:uuid;not null;index" json:"cari_id"`
	Type          string          `gorm:"type:varchar(20);not null" json:"type"`    // 'sales', 'purchase'
	Number        string          `gorm:"type:varchar(100);not null" json:"number"` // unique per company sequence-generated
	Date          time.Time       `gorm:"not null" json:"date"`
	DueDate       time.Time       `gorm:"not null" json:"due_date"`
	Currency      string          `gorm:"type:varchar(10);not null;default:'TRY'" json:"currency"`
	ExchangeRate  decimal.Decimal `gorm:"type:numeric(65,10);not null;default:1" json:"exchange_rate"`
	Subtotal      decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"subtotal"`
	DiscountTotal decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"discount_total"`
	TaxTotal      decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"tax_total"`
	Total         decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"total"`
	PaidTotal     decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"paid_total"`
	Status        string          `gorm:"type:varchar(50);not null;default:'draft'" json:"status"` // 'draft', 'sent', 'disputed', 'partial', 'paid', 'canceled'
	// ProcessStock false ise bu fatura stok hareketi YARATMAZ (hizmet/gider faturaları
	// gibi). Varsayılan true; iptal de bu bayrağa uyar (stok işlemeyen iptal de işlemez).
	ProcessStock  bool            `gorm:"not null;default:true" json:"process_stock"`
	Note          string          `gorm:"type:text" json:"note"`
	SentAt        *time.Time      `json:"sent_at"`
	LastSentAt    *time.Time      `json:"last_sent_at"`
	DisputeNote   string          `gorm:"type:text" json:"dispute_note"`
	DisputedAt    *time.Time      `json:"disputed_at"`
	PublicToken   string          `gorm:"type:varchar(64);uniqueIndex" json:"-"`
	Items         []InvoiceItem   `gorm:"foreignKey:InvoiceID;constraint:OnDelete:CASCADE" json:"items"`
	CreatedBy     *uuid.UUID      `gorm:"type:uuid" json:"created_by"`
	UpdatedBy     *uuid.UUID      `gorm:"type:uuid" json:"updated_by"`
	CreatedByUser *User           `gorm:"foreignKey:CreatedBy;constraint:OnDelete:SET NULL" json:"created_by_user,omitempty"`
	UpdatedByUser *User           `gorm:"foreignKey:UpdatedBy;constraint:OnDelete:SET NULL" json:"updated_by_user,omitempty"`
	CreatedAt     time.Time       `json:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at"`
}

type InvoiceItem struct {
	ID           uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID    uuid.UUID       `gorm:"type:uuid;not null;index" json:"company_id"`
	InvoiceID    uuid.UUID       `gorm:"type:uuid;not null;index" json:"invoice_id"`
	ProductID    *uuid.UUID      `gorm:"type:uuid" json:"product_id"` // nullable->products (for Module 7)
	Description  string          `gorm:"type:text" json:"description"`
	Quantity     decimal.Decimal `gorm:"type:numeric(65,4);not null;default:1" json:"quantity"`
	Unit         string          `gorm:"type:varchar(50)" json:"unit"`
	UnitPrice    decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"unit_price"`
	DiscountRate decimal.Decimal `gorm:"type:numeric(5,2);not null;default:0" json:"discount_rate"`
	TaxRate      decimal.Decimal `gorm:"type:numeric(5,2);not null;default:0" json:"tax_rate"` // e.g. 20 for 20%
	LineTotal    decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"line_total"` // satırın kendi dövizinde
	Currency       string          `gorm:"type:varchar(10);not null;default:'TRY'" json:"currency"`
	ExchangeRate   decimal.Decimal `gorm:"type:numeric(65,10);not null;default:1" json:"exchange_rate"`
	ExchangeRateOp string          `gorm:"type:varchar(5);not null;default:'*'" json:"exchange_rate_op"` // '*' veya '/'
	CreatedAt    time.Time       `json:"created_at"`
	UpdatedAt    time.Time       `json:"updated_at"`
}
